### PR TITLE
Adds HydrostaticSphericalCoriolis

### PR DESCRIPTION
**Pull request by glwagner**
~~This PR adapts `BetaPlane` coriolis implementation for curvilinear grids (following #1371 and [MITgcm docs](https://mitgcm.readthedocs.io/en/latest/algorithm/algorithm.html#id8)).~~

~~Note that the MITgcm docs specify averaging the Coriolis parameter / planetary vorticity (from corner / vorticity nodes, ffc, to the velocity nodes fcc and cfc). However, because the variation in the planetary vorticity is linear, I think the implementation in this PR is correct for beta planes? @jm-c please advise.~~

I would ~~also~~ like to add a Coriolis term valid for the sphere and hydrostatic models (which amounts to the "traditional approximation"). ~~I'll do this once #1380 is merged.~~ This PR depends on #1380.

Some feedback on what to call this object is certainly welcome! I have preliminary proposed "`HydrostaticSphericalCoriolis` because it was the first silly thing I could think of.

To do:

- [x] Add `HydrostaticSphericalCoriolis` plus unit tests

### CODE REVIEW

**Comment by francispoulin** [comment]
I will let @jm-c approve this one but wanted to say I looked ti over and it looks good to me.

As for the name, when  you say spherical Coriolis, I thought you meant the Coriolis parameter that depends on spherical coordinates, which is not the case.  Thsi clearly is a linear, hydrostatic beta-plane that works in curvilinear coordinates, if I understand this correctly.  Would `HydrostaticCurvilinearCoriolis` be appropriate?

**Review by jm-c — COMMENTED** [review]
I am not sure that I am following everything there. Is it for flux-form of vector-invariant ? otherwise, I don't recognize the energy-conserving discretization (thought it would have been nice to have).

**Comment by glwagner** [comment]
> As for the name, when you say spherical Coriolis, I thought you meant the Coriolis parameter that depends on spherical coordinates, which is not the case. Thsi clearly is a linear, hydrostatic beta-plane that works in curvilinear coordinates, if I understand this correctly. Would HydrostaticCurvilinearCoriolis be appropriate?

I haven't added `HydrostaticCurvilinearCoriolis` yet --- I'm planning to implement that once the PR is merged. At the current moment, I've only changed `BetaPlane`.

**Comment by glwagner** [comment]
> I am not sure that I am following everything there. Is it for flux-form of vector-invariant ? otherwise, I don't recognize the energy-conserving discretization (thought it would have been nice to have).

I didn't realize the the Coriolis implementation is different for the "flux-form" momentum equations than for vector-invariant momentum equations --- thanks for pointing that out @jm-c. I don't quite understand why this is the case. However, it is not that difficult to extend our Coriolis implementation to dispatch on the `advection` "scheme" / form if we need to.

After looking into the details (see below), it would appear that the implementation we are using is correct if

1. The horizontal grid is regular
2. The horizontal grid is curvilinear and we are using `momentum_advection=VectorInvariant()`

Because these are the only two possible cases, I believe the current Coriolis implementation is sufficient. However when/if we support curvilinear horizontal grids with flux form advection terms, we will (if we follow the MITgcm implementation) need to dispatch the evaluation of the Coriolis term on `momentum_advection` scheme.

Another question is whether it even makes sense / is physically consistent to support `FPlane` and `BetaPlane` on grids that are sectors of the sphere. Perhaps there's no point at all in implementing a curvilinear-friendly version of `FPlane` and `BetaPlane`. If this is true, we can revert back to the prior implementation of both for Rectilinear grids, and focus on a energy/enstrophy - conserving Coriolis scheme for curvilinear grids on the surface of the sphere.

# Some notes on Coriolis implementations for vector-invariant and flux form momentum schemes 

Focusing on the x-component, the [MITgcm documentation](https://mitgcm.readthedocs.io/en/latest/algorithm/algorithm.html#id8) writes that with a vector-invariant advection scheme, the Coriolis term that preserve potential enstrophy is

image

Note that all `h_* = 1` for the purposes of our discussion, and that in Oceananigans notation we have `Δx_g = Δxᶜᶠᶜ` (the grid length at the location of the y-velocity) and `Δx_c = Δxᶠᶜᶜ`.

For a flux-form advection scheme, the [MITgcm documentation](https://mitgcm.readthedocs.io/en/latest/algorithm/algorithm.html#coriolis-terms) states that for a hydrostatic implementation we would have

image

If the horizontal and vertical grids are separable, then `Δr_f = Δzᵃᵃᶜ` is independent of horizontal position and drops out of the above equation.

As a result, the current Coriolis implementation is not valid for a flux form advection scheme on a curvilinear grid.

**Comment by glwagner** [comment]
@jm-c please focus on this file for your review: https://github.com/CliMA/Oceananigans.jl/pull/1384/files#diff-fdb33e0cf3440fd4dd7498f8562a575415cd0adc4a4ea8e97f4cbe2d189810e1

File: src/Operators/Operators.jl [hunk 1]
```diff
@@ -26,12 +26,14 @@ export
 ##### Convinient aliases
 #####
 
-using Oceananigans.Grids: AbstractGrid, RegularCartesianGrid
-using Oceananigans.Grids: VerticallyStretchedCartesianGrid, AbstractRectilinearGrid
+using Oceananigans.Grids: AbstractGrid, RegularCartesianGrid, VerticallyStretchedCartesianGrid, RegularLatitudeLongitudeGrid
+using Oceananigans.Grids: AbstractGrid, AbstractRectilinearGrid, AbstractCurvilinearGrid, AbstractHorizontallyCurvilinearGrid
```
Comment (francispoulin) [inline src/Operators/Operators.jl:30]: We have AbstractGrid on lines 29 and 30. I presume we can delete one?

**Review by francispoulin — COMMENTED** [review]
I am not going to approve as I don't think I can vouche for the accuracy of all the equations but I have some very minor comments.

File: src/Coriolis/hydrostatic_spherical_coriolis.jl [hunk 2]
```diff
@@ -0,0 +1,67 @@
+using Oceananigans.Grids: RegularLatitudeLongitudeGrid
+using Oceananigans.Operators: Δx_vᶜᶠᵃ, Δy_uᶠᶜᵃ, Δxᶠᶜᵃ, Δyᶜᶠᵃ
+
+struct VectorInvariantEnergyConserving end
+struct VectorInvariantEnstrophyConserving end
+# struct FluxFormEnergyConserving end
```
Comment (ali-ramadhan) [inline src/Coriolis/hydrostatic_spherical_coriolis.jl:6]:
```suggestion
```

File: src/Coriolis/hydrostatic_spherical_coriolis.jl [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+@inline x_f_cross_U(i, j, k, grid::RegularLatitudeLongitudeGrid, coriolis::VIEnergy, U) =
+    @inbounds + ℑyᵃᶜᵃ(i, j, k, grid, f_ℑx_vᶠᶠᵃ, coriolis, U[2]) / Δxᶠᶜᵃ(i, j, k, grid)
+
+@inline y_f_cross_U(i, j, k, grid::RegularLatitudeLongitudeGrid, coriolis::VIEnergy, U) =
+    @inbounds - ℑxᶜᵃᵃ(i, j, k, grid, f_ℑy_uᶠᶠᵃ, coriolis, U[1]) / Δyᶜᶠᵃ(i, j, k, grid)  
+
+#####
+##### Show
+#####
+Base.show(io::IO, hydrostatic_spherical_coriolis::HydrostaticSphericalCoriolis{FT}) where FT =
```
Comment (ali-ramadhan) [inline src/Coriolis/hydrostatic_spherical_coriolis.jl:65]:
```suggestion
#####

Base.show(io::IO, hydrostatic_spherical_coriolis::HydrostaticSphericalCoriolis{FT}) where FT =
```

**Review by ali-ramadhan — APPROVED** [review]
Code looks fine to me but will let @jm-c comment on implementation.

**Pull request merged**